streams/src/main/java/org/apache/kafka/streams/KeyValue.java

```diff
@@ -42,4 +42,21 @@ public KeyValue(K key, V value) {
     public String toString() {
         return "KeyValue(" + key + ", " + value + ")";
     }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other instanceof KeyValue) {
+            KeyValue otherKV = (KeyValue) other;
+
+            return key.equals(otherKV.key) && value.equals(otherKV.value);
```

Comment:
The `KeyValue` class allows null values for `key` and `value` (at least I didn't see input validations such as throwing IAE in its constructor when either key or value are null).  So we must guard against nulls / NPEs here.